ClassTypes/PowerShell/HyperV/check-hyperv-vm/Test-HypervVmConfig.ps1

```diff
@@ -0,0 +1,568 @@
+ 
+        foreach ($localUser in $addedLocalUsers) {
+            $localUserName = $localUser | Select-Object -ExpandProperty Name
+ 
+            if ($(Get-LocalAdminStatus -LocalUserName $localUserName) -or $(Get-HypervAdminStatus -LocalUserName $localUserName)) {
+                Write-Host "Verified user '$($localUserName)' has permissions to use Hyper-V."
+            }
+            else {
+                if ($PSCmdlet.ShouldContinue("User '$($localUserName)' can not use Hyper-V.  Add user to Hyper-V Administrators Group?", "Add user to Hyper-V Administrators Group?", [ref] $YesToAll, [ref] $NoToAll )) {          
+                    Add-LocalGroupMember -Group $(Get-HyperVAdminGroup) -Member $localUser
+                    Write-Host "$($localUserName) added to Hyper-V Administrators group."   
+                }  
+            }
+        }
+    }
+    Write-Host "Permission check completed.`n"
+
+    # --------------------- Checking HYPER-V VM settings ---------------------------------------------------
+   
+
+    if ($ConfigFilePath) {
+        $configs = Get-ConfigurationSettings -SettingsFilePath $ConfigFilePath
+    }
+    else {
+        $configs = New-Object ConfigurationInfo
+    }
+
+    $vms = @(Get-VM)
+
+    if ($vms.Count -gt 0) {
+
+        Write-Host "******************************"
+        Write-Host "*    Testing HYPER-V VMs     *"
+        Write-Host "******************************"
+
+
+        # Warn if any VMs are in saved state
+        $savedStateVMs = @($vms | Where-Object { $_.State -eq "Saved" })
+        if ($savedStateVMs) {
+            Write-Warning "Found VM(s) that are in a saved state. VM may fail to start if processor of the Azure VM changes.`n`t$($savedStateVMs -join '`n`t')"
+            if ($PSCmdlet.ShouldContinue("Found VM(s) that are in a saved state. VM may fail to start if processor of the Azure VM changes.`n`t$($savedStateVMs -join '`n`t').", "Start and ShutDown Hyper-V VMs?", [ref] $YesToAll, [ref] $NoToAll )) {
+                foreach ($vm in $savedStateVMs) {
+                    $vm | Start-VM -WarningAction Continue
+                    $vm | Stop-VM -Force -WarningAction Continue
+                    Write-Verbose "Hyper-V VM $($vm.VMName) started and stopped."
+                }
+            }
+        }
+
+        # Premptively stop all VMs.  Most settings require the VM to be stopped before the setting can be updated.
+        $runningStateVMs = @($vms | Where-Object { $_.State -eq "Running" })
+        if ($runningStateVMs) {
+            if ($PSCmdlet.ShouldContinue("Found VM(s) that are running. VMs must be shutdown before settings are updated.", "ShutDown Hyper-V VMs?", [ref] $YesToAll, [ref] $NoToAll )) {
+                foreach ($vm in $runningStateVMs) {
+                    $vm | Stop-VM -Force -WarningAction Continue
+                    Write-Verbose "Hyper-V VM $($vm.VMName) stopped."
+                }
+            }
+        }
+
+        foreach ($vm in $vms) {
```

Comment:
If the user picked "No" above on shutting down the running VMs, should we continue to try and apply settings on all the VMs?  Or should we re-query and apply settings to only non-running VMs?  Another option would be to end the script...

Reply:
If the use chooses No, the `Set-HypervVmProperty` will re-query to verify the Hyper-V VM is stopped.  If user chooses no again, then an error will be printed that the Hyper-V VM must be stopped before the property is updated.  `Set-HypervVmProperty` will only be called if a property needs to be changed.

It's possible for the user to choose "no" and if nothing needs to be updated, it's ok.